Class 8 Pandas as pd

In [5]:
import pandas as pd

 chr() and ord()
ASCII = string encoding format

unicode= UNICODE  is a subset of ASCHII

In [1]:
s= input("Enter a character to get unicode value")
print(ord(s))

Enter a character to get unicode valueP
80


In [2]:
s= int(input("Enter a string to get unicode value"))
print(chr(s))

Enter a string to get unicode value78
N


In [3]:
"""to remove unicode error usr r or s or \\ before address
pd.read_excel(r'Attribute DataSet.xlsx')
"""

"to remove unicode error usr r or s or \\ before address\npd.read_excel(r'Attribute DataSet.xlsx')\n"

In [6]:
df = pd.read_csv("desktop/Attribute_dataset.csv")

In [7]:
df

Dress_ID    Style    Price  Rating  Size  Season   NeckLine  \
0    1006032852     Sexy      Low     4.6     M  Summer     o-neck   
1    1212192089   Casual      Low     0.0     L  Summer     o-neck   
2    1190380701  vintage     High     0.0     L  Automn     o-neck   
3     966005983    Brief  Average     4.6     L  Spring     o-neck   
4     876339541     cute      Low     4.5     M  Summer     o-neck   
..          ...      ...      ...     ...   ...     ...        ...   
474   990559192    Brief  Average     4.7     M  winter     o-neck   
475   713391965   Casual      Low     4.7     M  Spring     o-neck   
476   532874347   Casual  Average     4.7     M  Summer     v-neck   
477   655464934   Casual  Average     4.6     L  winter  boat-neck   
478   919930954   Casual      Low     4.4  free  Summer     v-neck   

    SleeveLength       Material  FabricType  Decoration Pattern Type  \
0      sleevless            NaN     chiffon     ruffles       animal   
1          Petal     microfiber         NaN     ruffles       animal   
2           full       polyster         NaN         NaN        print   
3           full           silk     chiffon  embroidary        print   
4      butterfly  chiffonfabric     chiffon         bow          dot   
..           ...            ...         ...         ...          ...   
474   halfsleeve        acrylic     chiffon         NaN      striped   
475         full       polyster         NaN         NaN        solid   
476         full         cotton         NaN        lace        solid   
477    sleevless           silk  broadcloth    applique        print   
478        short         cotton    Corduroy        lace        solid   

     Recommendation  
0                 1  
1                 0  
2                 0  
3                 1  
4                 0  
..              ...  
474               0  
475               1  
476               1  
477               1  
478               0  

[479 rows x 13 columns]

In [8]:
type(df)

pandas.core.frame.DataFrame

pd.read_excel('Attribute DataSet.xlsx',sheet_name= 'Sheet2' )

pd.read_excel('Attribute DataSet.xlsx',sheet_name= 'Sheet3', header= None, names= ['A','B','C','D','E'] )

In [9]:
pd.read_csv('https://raw.githubusercontent.com/selva86/datasets/master/Smarket.csv')

Year   Lag1   Lag2   Lag3   Lag4   Lag5   Volume  Today Direction
0     2001  0.381 -0.192 -2.624 -1.055  5.010  1.19130  0.959        Up
1     2001  0.959  0.381 -0.192 -2.624 -1.055  1.29650  1.032        Up
2     2001  1.032  0.959  0.381 -0.192 -2.624  1.41120 -0.623      Down
3     2001 -0.623  1.032  0.959  0.381 -0.192  1.27600  0.614        Up
4     2001  0.614 -0.623  1.032  0.959  0.381  1.20570  0.213        Up
...    ...    ...    ...    ...    ...    ...      ...    ...       ...
1245  2005  0.422  0.252 -0.024 -0.584 -0.285  1.88850  0.043        Up
1246  2005  0.043  0.422  0.252 -0.024 -0.584  1.28581 -0.955      Down
1247  2005 -0.955  0.043  0.422  0.252 -0.024  1.54047  0.130        Up
1248  2005  0.130 -0.955  0.043  0.422  0.252  1.42236 -0.298      Down
1249  2005 -0.298  0.130 -0.955  0.043  0.422  1.38254 -0.489      Down

[1250 rows x 9 columns]

In [10]:

pd.read_csv('https://raw.githubusercontent.com/selva86/datasets/master/BCF.csv')

date  values.I2P  values.P2I  values.JPI
0    1996-04-01   -2.580556   -0.836444        63.3
1    1996-05-01   -2.882258   -1.972581        66.3
2    1996-06-01   -1.742778   -0.478778        59.7
3    1996-07-01   -0.397796   -1.574301        63.3
4    1996-08-01   -0.977151   -1.797312        63.8
..          ...         ...         ...         ...
112  2005-08-01   -2.144935   -1.646989        49.0
113  2005-09-01   -3.543270   -1.769889        49.9
114  2005-10-01   -4.500484   -2.309946        50.6
115  2005-11-01   -4.098556   -1.528611        57.4
116  2005-12-01   -5.797043   -2.872849        52.9

[117 rows x 4 columns]

In [11]:
a = pd.read_html('https://www.basketball-reference.com/leagues/NBA_2015_totals.html')

In [12]:
a

[      Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
 0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
 1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
 2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
 3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
 4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
 ..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
 670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
 671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
 672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
 673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
 674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   
 
      DRB  TRB

In [13]:
type(a)

list

In [14]:
len(a)

1

In [15]:
a[0]

Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   

     DRB  TRB  AST  STL BLK  TOV   PF   PTS  
0    222  301   68   27  22   60  147   398  
1     19   28   16   16   7   14   24    94  
2    324  523   66   38  86   99  222   537  
3     54   77   15    4   9    9   30    60  
4    220  247  129   41   7  116  167  1035  
..   ...  ...  ...  ...  ..  ...  ...   ...  
670  284  411  173  124  25  117  171  1071  
671  170  245  135   86  17   75  115   685  
672  114  166   38   38   8   42   56   386  
673  265  362  100   34  49   62  156   472  
674  319  465  113   18  52   76  205   833  

[675 rows x 30 columns]

In [16]:
js=pd.read_json('https://api.github.com/repos/pandas-dev/pandas/issues')

In [17]:
js

url  \
0   https://api.github.com/repos/pandas-dev/pandas...   
1   https://api.github.com/repos/pandas-dev/pandas...   
2   https://api.github.com/repos/pandas-dev/pandas...   
3   https://api.github.com/repos/pandas-dev/pandas...   
4   https://api.github.com/repos/pandas-dev/pandas...   
5   https://api.github.com/repos/pandas-dev/pandas...   
6   https://api.github.com/repos/pandas-dev/pandas...   
7   https://api.github.com/repos/pandas-dev/pandas...   
8   https://api.github.com/repos/pandas-dev/pandas...   
9   https://api.github.com/repos/pandas-dev/pandas...   
10  https://api.github.com/repos/pandas-dev/pandas...   
11  https://api.github.com/repos/pandas-dev/pandas...   
12  https://api.github.com/repos/pandas-dev/pandas...   
13  https://api.github.com/repos/pandas-dev/pandas...   
14  https://api.github.com/repos/pandas-dev/pandas...   
15  https://api.github.com/repos/pandas-dev/pandas...   
16  https://api.github.com/repos/pandas-dev/pandas...   
17  https://api.github.com/repos/pandas-dev/pandas...   
18  https://api.github.com/repos/pandas-dev/pandas...   
19  https://api.github.com/repos/pandas-dev/pandas...   
20  https://api.github.com/repos/pandas-dev/pandas...   
21  https://api.github.com/repos/pandas-dev/pandas...   
22  https://api.github.com/repos/pandas-dev/pandas...   
23  https://api.github.com/repos/pandas-dev/pandas...   
24  https://api.github.com/repos/pandas-dev/pandas...   
25  https://api.github.com/repos/pandas-dev/pandas...   
26  https://api.github.com/repos/pandas-dev/pandas...   
27  https://api.github.com/repos/pandas-dev/pandas...   
28  https://api.github.com/repos/pandas-dev/pandas...   
29  https://api.github.com/repos/pandas-dev/pandas...   

                                    repository_url  \
0   https://api.github.com/repos/pandas-dev/pandas   
1   https://api.github.com/repos/pandas-dev/pandas   
2   https://api.github.com/repos/pandas-dev/pandas   
3   https://api.github.com/repos/pandas-dev/pandas   
4   https://api.github.com/repos/pandas-dev/pandas   
5   https://api.github.com/repos/pandas-dev/pandas   
6   https://api.github.com/repos/pandas-dev/pandas   
7   https://api.github.com/repos/pandas-dev/pandas   
8   https://api.github.com/repos/pandas-dev/pandas   
9   https://api.github.com/repos/pandas-dev/pandas   
10  https://api.github.com/repos/pandas-dev/pandas   
11  https://api.github.com/repos/pandas-dev/pandas   
12  https://api.github.com/repos/pandas-dev/pandas   
13  https://api.github.com/repos/pandas-dev/pandas   
14  https://api.github.com/repos/pandas-dev/pandas   
15  https://api.github.com/repos/pandas-dev/pandas   
16  https://api.github.com/repos/pandas-dev/pandas   
17  https://api.github.com/repos/pandas-dev/pandas   
18  https://api.github.com/repos/pandas-dev/pandas   
19  https://api.github.com/repos/pandas-dev/pandas   
20  https://api.github.com/repos/pandas-dev/pandas   
21  https://api.github.com/repos/pandas-dev/pandas   
22  https://api.github.com/repos/pandas-dev/pandas   
23  https://api.github.com/repos/pandas-dev/pandas   
24  https://api.github.com/repos/pandas-dev/pandas   
25  https://api.github.com/repos/pandas-dev/pandas   
26  https://api.github.com/repos/pandas-dev/pandas   
27  https://api.github.com/repos/pandas-dev/pandas   
28  https://api.github.com/repos/pandas-dev/pandas   
29  https://api.github.com/repos/pandas-dev/pandas   

                                           labels_url  \
0   https://api.github.com/repos/pandas-dev/pandas...   
1   https://api.github.com/repos/pandas-dev/pandas...   
2   https://api.github.com/repos/pandas-dev/pandas...   
3   https://api.github.com/repos/pandas-dev/pandas...   
4   https://api.github.com/repos/pandas-dev/pandas...   
5   https://api.github.com/repos/pandas-dev/pandas...   
6   https://api.github.com/repos/pandas-dev/pandas...   
7   https://api.github.com/repos/pandas-dev/pandas...   
8   https://api.github.com/repos/pandas-dev/pandas...   
9   https://api.github.com/repos/pand

In [18]:
js.columns

Index(['url', 'repository_url', 'labels_url', 'comments_url', 'events_url',
       'html_url', 'id', 'node_id', 'number', 'title', 'user', 'labels',
       'state', 'locked', 'assignee', 'assignees', 'milestone', 'comments',
       'created_at', 'updated_at', 'closed_at', 'author_association',
       'active_lock_reason', 'draft', 'pull_request', 'body', 'reactions',
       'timeline_url', 'performed_via_github_app', 'state_reason'],
      dtype='object')

In [19]:
js['user']

0     {'login': 'datapythonista', 'id': 10058240, 'n...
1     {'login': 'rhshadrach', 'id': 45562402, 'node_...
2     {'login': 'jbrockmendel', 'id': 8078968, 'node...
3     {'login': 'mroeschke', 'id': 10647082, 'node_i...
4     {'login': 'lohmanndouglas', 'id': 9731599, 'no...
5     {'login': 'jbrockmendel', 'id': 8078968, 'node...
6     {'login': 'mroeschke', 'id': 10647082, 'node_i...
7     {'login': 'rwijtvliet', 'id': 4106013, 'node_i...
8     {'login': 'jbrockmendel', 'id': 8078968, 'node...
9     {'login': 'lithomas1', 'id': 47963215, 'node_i...
10    {'login': 'rhshadrach', 'id': 45562402, 'node_...
11    {'login': 'rhshadrach', 'id': 45562402, 'node_...
12    {'login': 'Th3nn3ss', 'id': 25160953, 'node_id...
13    {'login': 'jbrockmendel', 'id': 8078968, 'node...
14    {'login': 'rspocz', 'id': 22218858, 'node_id':...
15    {'login': 'codamuse', 'id': 116989963, 'node_i...
16    {'login': 'jbrockmendel', 'id': 8078968, 'node...
17    {'login': 'cdcadman', 'id': 51368516, 'nod

In [20]:
js['user'][0]

{'login': 'datapythonista',
 'id': 10058240,
 'node_id': 'MDQ6VXNlcjEwMDU4MjQw',
 'avatar_url': 'https://avatars.githubusercontent.com/u/10058240?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/datapythonista',
 'html_url': 'https://github.com/datapythonista',
 'followers_url': 'https://api.github.com/users/datapythonista/followers',
 'following_url': 'https://api.github.com/users/datapythonista/following{/other_user}',
 'gists_url': 'https://api.github.com/users/datapythonista/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/datapythonista/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/datapythonista/subscriptions',
 'organizations_url': 'https://api.github.com/users/datapythonista/orgs',
 'repos_url': 'https://api.github.com/users/datapythonista/repos',
 'events_url': 'https://api.github.com/users/datapythonista/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/datapythonista/received_events',
 'type'

In [21]:
df.to_json('test.json')

In [22]:
pwd

'C:\\Users\\ACER'

In [23]:
df.to_html("D:\\test1.html")

In [24]:
df.to_csv("D:\\test2.csv")

In [25]:
df.to_json("D:\\test.json")

In [26]:
import pandas as pd

In [27]:
df = pd.read_csv("desktop/Attribute_dataset.csv")

In [28]:
df = pd.read_csv("desktop/spotify_data.csv")

In [36]:
df = pd.read_csv("desktop/Addresses.csv")

In [34]:
import pandas as pd

In [35]:
df1 = pd.read_excel("desktop/superstore_sales.Xlsx")

In [32]:
df1.head()

order_id order_date  ship_date       ship_mode    customer_name  \
0     AG-2011-2040 2011-01-01 2011-01-06  Standard Class  Toby Braunhardt   
1    IN-2011-47883 2011-01-01 2011-01-08  Standard Class      Joseph Holt   
2     HU-2011-1220 2011-01-01 2011-01-05    Second Class    Annie Thurman   
3  IT-2011-3647632 2011-01-01 2011-01-05    Second Class     Eugene Moren   
4    IN-2011-47883 2011-01-01 2011-01-08  Standard Class      Joseph Holt   

       segment            state    country  market   region  ...  \
0     Consumer      Constantine    Algeria  Africa   Africa  ...   
1     Consumer  New South Wales  Australia    APAC  Oceania  ...   
2     Consumer         Budapest    Hungary    EMEA     EMEA  ...   
3  Home Office        Stockholm     Sweden      EU    North  ...   
4     Consumer  New South Wales  Australia    APAC  Oceania  ...   

          category sub_category                 product_name    sales  \
0  Office Supplies      Storage          Tenex Lockers, Blue  408.300   
1  Office Supplies     Supplies     Acme Trimmer, High Speed  120.366   
2  Office Supplies      Storage      Tenex Box, Single Width   66.120   
3  Office Supplies        Paper  Enermax Note Cards, Premium   44.865   
4        Furniture  Furnishings   Eldon Light Bulb, Duo Pack  113.670   

   quantity  discount   profit  shipping_cost  order_priority  year  
0         2       0.0  106.140          35.46          Medium  2011  
1         3       0.1   36.036           9.72          Medium  2011  
2         4       0.0   29.640           8.17            High  2011  
3         3       0.5  -26.055           4.82            High  2011  
4         5       0.1   37.770           4.70          Medium  2011  

[5 rows x 21 columns]

In [33]:
df1.tail()

order_id order_date  ship_date       ship_mode     customer_name  \
51285  CA-2014-115427 2014-12-31 2015-01-04  Standard Class        Erica Bern   
51286    MO-2014-2560 2014-12-31 2015-01-05  Standard Class         Liz Preis   
51287  MX-2014-110527 2014-12-31 2015-01-02    Second Class  Charlotte Melton   
51288  MX-2014-114783 2014-12-31 2015-01-06  Standard Class     Tamara Dahlen   
51289  CA-2014-156720 2014-12-31 2015-01-04  Standard Class     Jill Matthias   

         segment             state        country  market   region  ...  \
51285  Corporate        California  United States      US     West  ...   
51286   Consumer  Souss-Massa-Draâ        Morocco  Africa   Africa  ...   
51287   Consumer           Managua      Nicaragua   LATAM  Central  ...   
51288   Consumer         Chihuahua         Mexico   LATAM    North  ...   
51289   Consumer          Colorado  United States      US     West  ...   

              category sub_category  \
51285  Office Supplies      Binders   
51286  Office Supplies      Binders   
51287  Office Supplies       Labels   
51288  Office Supplies       Labels   
51289  Office Supplies    Fasteners   

                                          product_name   sales  quantity  \
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl  13.904         2   
51286          Wilson Jones Hole Reinforcements, Clear   3.990         1   
51287           Hon Color Coded Labels, 5000 Label Set  26.400         3   
51288           Hon Legal Exhibit Labels, Alphabetical   7.120         1   
51289                              Bagged Rubber Bands   3.024         3   

       discount   profit  shipping_cost  order_priority  year  
51285       0.2   4.5188          0.890          Medium  2014  
51286       0.0   0.4200          0.490          Medium  2014  
51287       0.0  12.3600          0.350          Medium  2014  
51288       0.0   0.5600          0.199          Medium  2014  
51289       0.2  -0.6048          0.170          Medium  2014  

[5 rows x 21 columns]

In [37]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [38]:
df1["order_date"]

0       2011-01-01
1       2011-01-01
2       2011-01-01
3       2011-01-01
4       2011-01-01
           ...    
51285   2014-12-31
51286   2014-12-31
51287   2014-12-31
51288   2014-12-31
51289   2014-12-31
Name: order_date, Length: 51290, dtype: datetime64[ns]

In [39]:
type(df1["order_date"])

pandas.core.series.Series

In [40]:
df1.dtypes

order_id                  object
order_date        datetime64[ns]
ship_date         datetime64[ns]
ship_mode                 object
customer_name             object
segment                   object
state                     object
country                   object
market                    object
region                    object
product_id                object
category                  object
sub_category              object
product_name              object
sales                    float64
quantity                   int64
discount                 float64
profit                   float64
shipping_cost            float64
order_priority            object
year                       int64
dtype: object

In [41]:
df1["profit"]

0        106.1400
1         36.0360
2         29.6400
3        -26.0550
4         37.7700
           ...   
51285      4.5188
51286      0.4200
51287     12.3600
51288      0.5600
51289     -0.6048
Name: profit, Length: 51290, dtype: float64

In [42]:
df1.dtypes.isnull()

order_id          False
order_date        False
ship_date         False
ship_mode         False
customer_name     False
segment           False
state             False
country           False
market            False
region            False
product_id        False
category          False
sub_category      False
product_name      False
sales             False
quantity          False
discount          False
profit            False
shipping_cost     False
order_priority    False
year              False
dtype: bool

In [43]:
type(df1["profit"])

pandas.core.series.Series

In [44]:
df1[["profit","year"]]

profit  year
0      106.1400  2011
1       36.0360  2011
2       29.6400  2011
3      -26.0550  2011
4       37.7700  2011
...         ...   ...
51285    4.5188  2014
51286    0.4200  2014
51287   12.3600  2014
51288    0.5600  2014
51289   -0.6048  2014

[51290 rows x 2 columns]

In [45]:
df1[["profit"]]

profit
0      106.1400
1       36.0360
2       29.6400
3      -26.0550
4       37.7700
...         ...
51285    4.5188
51286    0.4200
51287   12.3600
51288    0.5600
51289   -0.6048

[51290 rows x 1 columns]

In [46]:
df1[['order_id', 'order_date', 'ship_date', 'ship_mode']]

order_id order_date  ship_date       ship_mode
0         AG-2011-2040 2011-01-01 2011-01-06  Standard Class
1        IN-2011-47883 2011-01-01 2011-01-08  Standard Class
2         HU-2011-1220 2011-01-01 2011-01-05    Second Class
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class
4        IN-2011-47883 2011-01-01 2011-01-08  Standard Class
...                ...        ...        ...             ...
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class

[51290 rows x 4 columns]

In [47]:
df1.describe()  #analytics about dataframe

sales      quantity      discount        profit  shipping_cost  \
count  51290.000000  51290.000000  51290.000000  51290.000000   51290.000000   
mean     246.490581      3.476545      0.142908     28.641740      26.375818   
std      487.565361      2.278766      0.212280    174.424113      57.296810   
min        0.444000      1.000000      0.000000  -6599.978000       0.002000   
25%       30.758625      2.000000      0.000000      0.000000       2.610000   
50%       85.053000      3.000000      0.000000      9.240000       7.790000   
75%      251.053200      5.000000      0.200000     36.810000      24.450000   
max    22638.480000     14.000000      0.850000   8399.976000     933.570000   

               year  
count  51290.000000  
mean    2012.777208  
std        1.098931  
min     2011.000000  
25%     2012.000000  
50%     2013.000000  
75%     2014.000000  
max     2014.000000

1.no null value in above table as all count is same
2.Avg Qty ordered is 3.47. avg discount is 0.14, avg profit is 28.64
3.min and max qty ordered is 1 and 14.
4.min profit loss is 6599.978
5.25 percentile of data

In [48]:
# filter out column having datatype object

In [49]:
df1.dtypes == "object"

order_id           True
order_date        False
ship_date         False
ship_mode          True
customer_name      True
segment            True
state              True
country            True
market             True
region             True
product_id         True
category           True
sub_category       True
product_name       True
sales             False
quantity          False
discount          False
profit            False
shipping_cost     False
order_priority     True
year              False
dtype: bool

In [50]:
df1.dtypes[df1.dtypes == "object"]

order_id          object
ship_mode         object
customer_name     object
segment           object
state             object
country           object
market            object
region            object
product_id        object
category          object
sub_category      object
product_name      object
order_priority    object
dtype: object

In [51]:
type(df1.dtypes[df1.dtypes == "object"])

pandas.core.series.Series

In [52]:
df1.dtypes[df1.dtypes == "object"].index

Index(['order_id', 'ship_mode', 'customer_name', 'segment', 'state', 'country',
       'market', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'order_priority'],
      dtype='object')

In [53]:
df1[df1.dtypes[df1.dtypes == "object"].index]

order_id       ship_mode     customer_name      segment  \
0         AG-2011-2040  Standard Class   Toby Braunhardt     Consumer   
1        IN-2011-47883  Standard Class       Joseph Holt     Consumer   
2         HU-2011-1220    Second Class     Annie Thurman     Consumer   
3      IT-2011-3647632    Second Class      Eugene Moren  Home Office   
4        IN-2011-47883  Standard Class       Joseph Holt     Consumer   
...                ...             ...               ...          ...   
51285   CA-2014-115427  Standard Class        Erica Bern    Corporate   
51286     MO-2014-2560  Standard Class         Liz Preis     Consumer   
51287   MX-2014-110527    Second Class  Charlotte Melton     Consumer   
51288   MX-2014-114783  Standard Class     Tamara Dahlen     Consumer   
51289   CA-2014-156720  Standard Class     Jill Matthias     Consumer   

                  state        country  market   region        product_id  \
0           Constantine        Algeria  Africa   Africa  OFF-TEN-10000025   
1       New South Wales      Australia    APAC  Oceania   OFF-SU-10000618   
2              Budapest        Hungary    EMEA     EMEA  OFF-TEN-10001585   
3             Stockholm         Sweden      EU    North   OFF-PA-10001492   
4       New South Wales      Australia    APAC  Oceania   FUR-FU-10003447   
...                 ...            ...     ...      ...               ...   
51285        California  United States      US     West   OFF-BI-10002103   
51286  Souss-Massa-Draâ        Morocco  Africa   Africa  OFF-WIL-10001069   
51287           Managua      Nicaragua   LATAM  Central   OFF-LA-10004182   
51288         Chihuahua         Mexico   LATAM    North   OFF-LA-10000413   
51289          Colorado  United States      US     West   OFF-FA-10003472   

              category sub_category  \
0      Office Supplies      Storage   
1      Office Supplies     Supplies   
2      Office Supplies      Storage   
3      Office Supplies        Paper   
4            Furniture  Furnishings   
...                ...          ...   
51285  Office Supplies      Binders   
51286  Office Supplies      Binders   
51287  Office Supplies       Labels   
51288  Office Supplies       Labels   
51289  Office Supplies    Fasteners   

                                          product_name order_priority  
0                                  Tenex Lockers, Blue         Medium  
1                             Acme Trimmer, High Speed         Medium  
2                              Tenex Box, Single Width           High  
3                          Enermax Note Cards, Premium           High  
4                           Eldon Light Bulb, Duo Pack         Medium  
...                                                ...            ...  
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl         Medium  
51286          Wilson Jones Hole Reinforcements, Clear         Medium  
51287           Hon Color Coded Labels, 5000 Label Set         Medium  
51288           Hon Legal Exhibit Labels, Alphabetical         Medium  
51289                              Bagged Rubber Bands         Medium  

[51290 rows x 13 columns]

In [54]:
df1[df1.dtypes[df1.dtypes == "object"].index].describe()

order_id       ship_mode    customer_name   segment       state  \
count            51290           51290            51290     51290       51290   
unique           25035               4              795         3        1094   
top     CA-2014-100111  Standard Class  Muhammed Yedwab  Consumer  California   
freq                14           30775              108     26518        2001   

              country market   region       product_id         category  \
count           51290  51290    51290            51290            51290   
unique            147      7       13            10292                3   
top     United States   APAC  Central  OFF-AR-10003651  Office Supplies   
freq             9994  11002    11117               35            31273   

       sub_category product_name order_priority  
count         51290        51290          51290  
unique           17         3788              4  
top         Binders      Staples         Medium  
freq           6152          227          29433

1.no null vale in this dataframe
2.25035 are unique orders
3.total 4 shipment modes
4.total 795 customers only

In [55]:
df1.dtypes[df1.dtypes == "float64"]

sales            float64
discount         float64
profit           float64
shipping_cost    float64
dtype: object

In [56]:
df1.dtypes[df1.dtypes == "int64"]

quantity    int64
year        int64
dtype: object

In [57]:
df1.dtypes[df1.dtypes == "float64"].index

Index(['sales', 'discount', 'profit', 'shipping_cost'], dtype='object')

In [58]:
df1[df1.dtypes[df1.dtypes == "float64"].index]

sales  discount    profit  shipping_cost
0      408.300       0.0  106.1400         35.460
1      120.366       0.1   36.0360          9.720
2       66.120       0.0   29.6400          8.170
3       44.865       0.5  -26.0550          4.820
4      113.670       0.1   37.7700          4.700
...        ...       ...       ...            ...
51285   13.904       0.2    4.5188          0.890
51286    3.990       0.0    0.4200          0.490
51287   26.400       0.0   12.3600          0.350
51288    7.120       0.0    0.5600          0.199
51289    3.024       0.2   -0.6048          0.170

[51290 rows x 4 columns]

In [59]:
df1[df1.dtypes[df1.dtypes == "int64"].index]

quantity  year
0             2  2011
1             3  2011
2             4  2011
3             3  2011
4             5  2011
...         ...   ...
51285         2  2014
51286         1  2014
51287         3  2014
51288         1  2014
51289         3  2014

[51290 rows x 2 columns]

In [60]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [61]:
df1["order_id"]

0           AG-2011-2040
1          IN-2011-47883
2           HU-2011-1220
3        IT-2011-3647632
4          IN-2011-47883
              ...       
51285     CA-2014-115427
51286       MO-2014-2560
51287     MX-2014-110527
51288     MX-2014-114783
51289     CA-2014-156720
Name: order_id, Length: 51290, dtype: object

In [62]:
df1["order_id"][1:4:2]

1      IN-2011-47883
3    IT-2011-3647632
Name: order_id, dtype: object

In [63]:
df1["order_id"][1:40:2]

1       IN-2011-47883
3     IT-2011-3647632
5       IN-2011-47883
7       IN-2011-79397
9        IZ-2011-4680
11      IN-2011-65159
13      IN-2011-33652
15     MX-2011-160234
17      ID-2011-80230
19      ID-2011-12596
21       IR-2011-7690
23       TZ-2011-7370
25      IN-2011-65159
27     MX-2011-111255
29     MX-2011-158771
31     MX-2011-111255
33     US-2011-136007
35     MX-2011-159373
37    IT-2011-2942451
39      IN-2011-33036
Name: order_id, dtype: object

In [64]:
df1["order_id"][1:40:-1]

Series([], Name: order_id, dtype: object)

In [65]:
df1["order_id"][40:1:-4]

40    IT-2011-2942451
36     MX-2011-159373
32     MX-2011-140641
28     MX-2011-140641
24       IZ-2011-4680
20      IN-2011-79397
16        IR-2011-770
12    ES-2011-4869686
8       ID-2011-80230
4       IN-2011-47883
Name: order_id, dtype: object

In [66]:
df1["order_id"][::-4]

51289    CA-2014-156720
51285    CA-2014-115427
51281    CA-2014-115427
51277    MX-2014-141712
51273    MX-2014-114783
              ...      
17        ID-2011-80230
13        IN-2011-33652
9          IZ-2011-4680
5         IN-2011-47883
1         IN-2011-47883
Name: order_id, Length: 12823, dtype: object

In [67]:
df1["order_id"][-30:-40:-4]

51260    CA-2014-143259
51256    CA-2014-143259
51252      OD-2014-9490
Name: order_id, dtype: object

In [68]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year'],
      dtype='object')

In [69]:
df1["catagory1"] = "nilam"

In [70]:
df1.pop("catagory1")

0        nilam
1        nilam
2        nilam
3        nilam
4        nilam
         ...  
51285    nilam
51286    nilam
51287    nilam
51288    nilam
51289    nilam
Name: catagory1, Length: 51290, dtype: object

In [71]:
df1

order_id order_date  ship_date       ship_mode  \
0         AG-2011-2040 2011-01-01 2011-01-06  Standard Class   
1        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
2         HU-2011-1220 2011-01-01 2011-01-05    Second Class   
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class   
4        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
...                ...        ...        ...             ...   
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class   
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class   
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class   
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class   
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draâ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ...         category sub_category  \
0       Africa  ...  Office Supplies      Storage   
1      Oceania  ...  Office Supplies     Supplies   
2         EMEA  ...  Office Supplies      Storage   
3        North  ...  Office Supplies        Paper   
4      Oceania  ...        Furniture  Furnishings   
...        ...  ...              ...          ...   
51285     West  ...  Office Supplies      Binders   
51286   Africa  ...  Office Supplies      Binders   
51287  Central  ...  Office Supplies       Labels   
51288    North  ...  Office Supplies       Labels   
51289     West  ...  Office Supplies    Fasteners   

                                          product_name    sales  quantity  \
0                                  Tenex Lockers, Blue  408.300         2   
1                             Acme Trimmer, High Speed  120.366         3   
2                              Tenex Box, Single Width   66.120         4   
3                          Enermax Note Cards, Premium   44.865         3   
4                           Eldon Light Bulb, Duo Pack  113.670         5   
...                                                ...      ...       ...   
51285  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl   13.904         2   
51286          Wilson Jones Hole Reinforcements, Clear    3.990         1   
51287           Hon Color Coded Labels, 5000 Label Set   26.400         3   
51288           Hon Legal Exhibit Labels, Alphabetical    7.120         1   
51289                              Bagged Rubber Bands    3.024         3   

       discount    profit  shipping_cost  order_priority  year  
0           0.0  106.1400         35.460          Medium  2011  
1           0.1   36.0360          9.720          Medium  2011  
2           0.0   29.6400          8.170            High  2011  
3           0.5  -26.0550          4.820            High  2011  
4           0.1   37.7700          4.700          Medium  2011  
...         ...       ...            ...             ...   ...  
51285       0.2    4.5188          0.890          Medium  2014  
51286       0.0    0.4200          0.490          Medium  2014  
51287       0.0   12.3600          0.350          Medium  2014  
51288       0.0    0.5600          0.199          Medium  20

In [72]:
df1["category1"] = "nilam"

In [73]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1'],
      dtype='object')

In [74]:
df1.columns.isnull()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False])

In [75]:
# person name who has received maximum profit

In [76]:
df1["profit"] == max(df1["profit"])

0        False
1        False
2        False
3        False
4        False
         ...  
51285    False
51286    False
51287    False
51288    False
51289    False
Name: profit, Length: 51290, dtype: bool

In [77]:
df1[df1["profit"] == max(df1["profit"])]

order_id order_date  ship_date       ship_mode customer_name  \
29530  CA-2013-118689 2013-10-03 2013-10-10  Standard Class  Tamara Chand   

         segment    state        country market   region  ... sub_category  \
29530  Corporate  Indiana  United States     US  Central  ...      Copiers   

                                product_name     sales quantity  discount  \
29530  Canon imageCLASS 2200 Advanced Copier  17499.95        5       0.0   

         profit  shipping_cost  order_priority  year category1  
29530  8399.976         349.07          Medium  2013     nilam  

[1 rows x 22 columns]

In [78]:
df1[df1["profit"] == max(df1["profit"])]["customer_name"]

29530    Tamara Chand
Name: customer_name, dtype: object

In [79]:
df1[df1["country"] == "Sweden"]

order_id order_date  ship_date       ship_mode    customer_name  \
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class     Eugene Moren   
168    IT-2011-5951216 2011-01-13 2011-01-20  Standard Class   Stuart Calhoun   
477    IT-2011-5645183 2011-02-03 2011-02-05    Second Class       Max Ludwig   
1342   IT-2011-5233011 2011-03-31 2011-04-06  Standard Class     Todd Sumrall   
1554   IT-2011-2419493 2011-04-12 2011-04-13     First Class   Alice McCarthy   
...                ...        ...        ...             ...              ...   
50948  IT-2014-5217096 2014-12-26 2014-12-31  Standard Class     Trudy Glocke   
50949  IT-2014-5217096 2014-12-26 2014-12-31  Standard Class     Trudy Glocke   
50972  IT-2014-4046179 2014-12-27 2014-12-30    Second Class     Jeremy Farry   
51086  IT-2014-2362249 2014-12-29 2015-01-01     First Class  Mitch Gastineau   
51121  IT-2014-2362249 2014-12-29 2015-01-01     First Class  Mitch Gastineau   

           segment            state country market region  ... sub_category  \
3      Home Office        Stockholm  Sweden     EU  North  ...        Paper   
168       Consumer  Västra Götaland  Sweden     EU  North  ...          Art   
477    Home Office          Uppsala  Sweden     EU  North  ...          Art   
1342     Corporate  Västra Götaland  Sweden     EU  North  ...  Furnishings   
1554     Corporate        Stockholm  Sweden     EU  North  ...       Phones   
...            ...              ...     ...    ...    ...  ...          ...   
50948     Consumer        Stockholm  Sweden     EU  North  ...  Accessories   
50949     Consumer        Stockholm  Sweden     EU  North  ...  Furnishings   
50972     Consumer        Stockholm  Sweden     EU  North  ...   Appliances   
51086    Corporate        Stockholm  Sweden     EU  North  ...          Art   
51121    Corporate        Stockholm  Sweden     EU  North  ...          Art   

                              product_name     sales quantity  discount  \
3              Enermax Note Cards, Premium    44.865        3       0.5   
168              Sanford Pens, Fluorescent    30.900        5       0.5   
477    Boston Pencil Sharpener, Easy-Erase    90.630        6       0.5   
1342         Advantus Photo Frame, Durable    40.008        2       0.6   
1554                   Nokia Headset, VoIP    73.200        2       0.5   
...                                    ...       ...      ...       ...   
50948        Belkin Flash Drive, Bluetooth    20.970        1       0.5   
50949        Advantus Photo Frame, Durable    20.004        1       0.6   
50972       KitchenAid Refrigerator, Black  1582.110        6       0.5   
51086                BIC Pens, Fluorescent    31.080        4       0.5   
51121           Sanford Highlighters, Blue     7.875        1       0.5   

        profit  shipping_cost  order_priority  year category1  
3      -26.055           4.82            High  2011     nilam  
168    -13.650           2.16          Medium  2011     nilam  
477    -61.650           6.64          Medium  2011     nilam  
1342   -44.052           2.78          Medium  2011     nilam  
1554   -11.760          22.18            High  2011     nilam  
...        ...            ...             ...   ...       ...  
50948   -5.460           0.98          Medium  2014     nilam  
50949  -22.026           0.98          Medium  2014     nilam  
50972 -443.070          55.39          Medium  2014     nilam  
51086  -10.680           8.44            High  2014     nilam  
51121   -6.165           2.00            High  2014     nilam  

[203 rows x 22 columns]

In [80]:
len(df1[df1["country"] == "Sweden"])

203

In [81]:
df1[df1["shipping_cost"] > 80]

order_id order_date  ship_date       ship_mode  \
7       IN-2011-79397 2011-01-03 2011-01-03        Same Day   
8       ID-2011-80230 2011-01-03 2011-01-09  Standard Class   
9        IZ-2011-4680 2011-01-03 2011-01-07  Standard Class   
27     MX-2011-111255 2011-01-04 2011-01-09    Second Class   
43     MX-2011-109267 2011-01-05 2011-01-09  Standard Class   
...               ...        ...        ...             ...   
51154   IN-2014-49206 2014-12-30 2015-01-01    Second Class   
51228  MX-2014-116267 2014-12-31 2015-01-03    Second Class   
51229   IN-2014-43550 2014-12-31 2015-01-01     First Class   
51230    RS-2014-1460 2014-12-31 2015-01-02    Second Class   
51231   IN-2014-30390 2014-12-31 2015-01-03     First Class   

             customer_name    segment            state      country market  \
7              Kean Nguyen  Corporate  New South Wales    Australia   APAC   
8             Ken Lonsdale   Consumer         Auckland  New Zealand   APAC   
9         Lindsay Williams  Corporate           Ninawa         Iraq   EMEA   
27       Russell Applegate   Consumer           Parana       Brazil  LATAM   
43       Jennifer Halladay   Consumer         Veracruz       Mexico  LATAM   
...                    ...        ...              ...          ...    ...   
51154          Toby Ritter   Consumer       Jawa Barat    Indonesia   APAC   
51228           Erica Bern  Corporate        São Paulo       Brazil  LATAM   
51229  Marina Lichtenstein  Corporate          Jakarta    Indonesia   APAC   
51230         Peter Bühler   Consumer    Bashkortostan       Russia   EMEA   
51231     Justin Deggeller  Corporate          Bangkok     Thailand   APAC   

               region  ... sub_category  \
7             Oceania  ...   Appliances   
8             Oceania  ...      Copiers   
9                EMEA  ...       Chairs   
27              South  ...    Bookcases   
43              North  ...   Appliances   
...               ...  ...          ...   
51154  Southeast Asia  ...      Storage   
51228           South  ...      Copiers   
51229  Southeast Asia  ...    Bookcases   
51230            EMEA  ...      Copiers   
51231  Southeast Asia  ...       Tables   

                                  product_name      sales quantity  discount  \
7                      Hoover Microwave, White   276.1020        1     0.100   
8                  Hewlett Wireless Fax, Laser   912.4560        4     0.400   
9             Novimex Swivel Stool, Set of Two   666.8400        4     0.000   
27                Dania Classic Bookcase, Pine  1648.4400        6     0.000   
43                         Hoover Stove, Black  3028.8000        8     0.000   
...                                        ...        ...      ...       ...   
51154              Rogers Shelving, Wire Frame   457.1640        9     0.170   
51228                Canon Wireless Fax, Color  1264.4660        5     0.002   
51229           Safco Library with Doors, Pine  1091.2806        3     0.070   
51230           Hewlett Personal Copier, Laser   865.6200        6     0.000   
51231  Lesro Training Table, Adjustable Height  1048.7313        9     0.570   

          profit  shipping_cost  order_priority  year category1  
7       110.4120        125.320        Critical  2011     nilam  
8      -319.4640        107.100             Low  2011     nilam  
9       253.3200         81.260            High  2011     nilam  
27      609.8400        109.129          Medium  2011     nilam  
43      999.3600        191.201          Medium  2011     nilam  
...          ...            ...             ...   ...       ...  
51154    65.9340        101.880            High  2014     nilam  
51228   301.4660        253.254            High  2014     nilam  
51229    46.9206        243.110            High  2014     nilam  
51230    51.8400        138.180            High  2014     nilam  
51231 -1195.2387         86.860            High  2014     nilam  

[3984 rows x 22 columns]

In [82]:
len(df1[df1["shipping_cost"] > 80])

3984

In [83]:
len(df1[df1["shipping_cost"] > 20])

14844

In [84]:
df1[df1["shipping_cost"] >80]["country"]

7          Australia
8        New Zealand
9               Iraq
27            Brazil
43            Mexico
            ...     
51154      Indonesia
51228         Brazil
51229      Indonesia
51230         Russia
51231       Thailand
Name: country, Length: 3984, dtype: object

In [85]:
df1[(df1["shipping_cost"] >80) & (df1["profit"] <10)]

order_id order_date  ship_date       ship_mode  \
8        ID-2011-80230 2011-01-03 2011-01-09  Standard Class   
200    ES-2011-1374122 2011-01-15 2011-01-18     First Class   
257     MX-2011-146584 2011-01-21 2011-01-27  Standard Class   
341      ID-2011-36011 2011-01-26 2011-01-27     First Class   
451    ES-2011-1640672 2011-02-02 2011-02-05     First Class   
...                ...        ...        ...             ...   
50970  ES-2014-4296634 2014-12-27 2015-01-02  Standard Class   
51034    ID-2014-68183 2014-12-29 2014-12-31     First Class   
51037  IT-2014-2245117 2014-12-29 2014-12-31     First Class   
51153    ID-2014-51509 2014-12-30 2015-01-01    Second Class   
51231    IN-2014-30390 2014-12-31 2015-01-03     First Class   

          customer_name      segment                         state  \
8          Ken Lonsdale     Consumer                      Auckland   
200        Craig Leslie  Home Office            Nord-Pas-de-Calais   
257        Hunter Lopez     Consumer                       Managua   
341    Heather Kirkland    Corporate                    Jawa Barat   
451          Ken Heidel    Corporate                        Saxony   
...                 ...          ...                           ...   
50970       Harry Marie    Corporate                      Brittany   
51034         Adam Hart    Corporate                      Victoria   
51037        Sarah Bern     Consumer                 North Holland   
51153      Naresj Patel     Consumer  Australian Capital Territory   
51231  Justin Deggeller    Corporate                       Bangkok   

           country market          region  ... sub_category  \
8      New Zealand   APAC         Oceania  ...      Copiers   
200         France     EU         Central  ...    Bookcases   
257      Nicaragua  LATAM         Central  ...       Tables   
341      Indonesia   APAC  Southeast Asia  ...       Chairs   
451        Germany     EU         Central  ...       Phones   
...            ...    ...             ...  ...          ...   
50970       France     EU         Central  ...   Appliances   
51034    Australia   APAC         Oceania  ...   Appliances   
51037  Netherlands     EU         Central  ...       Chairs   
51153    Australia   APAC         Oceania  ...       Chairs   
51231     Thailand   APAC  Southeast Asia  ...       Tables   

                                       product_name      sales quantity  \
8                       Hewlett Wireless Fax, Laser   912.4560        4   
200           Dania Library with Doors, Traditional   652.3200        2   
257    Chromcraft Conference Table, Fully Assembled  2797.2480        6   
341       Novimex Executive Leather Armchair, Black   334.3035        1   
451               Motorola Signal Booster, Cordless   583.9200        8   
...                                             ...        ...      ...   
50970               Hamilton Beach Microwave, White  1010.3400        4   
51034                             Hoover Stove, Red  1534.8690        3   
51037                  Novimex Rocking Chair, Black   516.0000        8   
51153        Harbour Creations Rocking Chair, Black   433.9800        5   
51231       Lesro Training Table, Adjustable Height  1048.7313        9   

       discount     profit  shipping_cost  order_priority  year category1  
8          0.40  -319.4640        107.100             Low  2011     nilam  
200        0.10     7.2000         88.520          Medium  2011     nilam  
257        0.20    -0.0720        320.376             Low  2011     nilam  
341        0.27  -109.9365         92.670            High  2011     nilam  
451        0.50   -58.5600        109.140            High  2011     nilam  
...         ...        ...            ...             ...   ...       ...  
50970      0.10   -44.9400        108.270             Low  2014     nilam  
51034      0.10  -102.4110        255.420          Medium  2014     nilam  
51037      0.50  -392.1600         91.670        Critical  201

In [86]:
date = pd.to_datetime(df1["order_date"])

In [87]:
type(date[0])

pandas._libs.tslibs.timestamps.Timestamp

In [88]:
df1["converted_order_date"] = date

In [89]:
df1

order_id order_date  ship_date       ship_mode  \
0         AG-2011-2040 2011-01-01 2011-01-06  Standard Class   
1        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
2         HU-2011-1220 2011-01-01 2011-01-05    Second Class   
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class   
4        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
...                ...        ...        ...             ...   
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class   
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class   
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class   
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class   
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draâ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ...                                     product_name    sales  \
0       Africa  ...                              Tenex Lockers, Blue  408.300   
1      Oceania  ...                         Acme Trimmer, High Speed  120.366   
2         EMEA  ...                          Tenex Box, Single Width   66.120   
3        North  ...                      Enermax Note Cards, Premium   44.865   
4      Oceania  ...                       Eldon Light Bulb, Duo Pack  113.670   
...        ...  ...                                              ...      ...   
51285     West  ...  Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl   13.904   
51286   Africa  ...          Wilson Jones Hole Reinforcements, Clear    3.990   
51287  Central  ...           Hon Color Coded Labels, 5000 Label Set   26.400   
51288    North  ...           Hon Legal Exhibit Labels, Alphabetical    7.120   
51289     West  ...                              Bagged Rubber Bands    3.024   

      quantity discount    profit  shipping_cost  order_priority  year  \
0            2      0.0  106.1400         35.460          Medium  2011   
1            3      0.1   36.0360          9.720          Medium  2011   
2            4      0.0   29.6400          8.170            High  2011   
3            3      0.5  -26.0550          4.820            High  2011   
4            5      0.1   37.7700          4.700          Medium  2011   
...        ...      ...       ...            ...             ...   ...   
51285        2      0.2    4.5188          0.890          Medium  2014   
51286        1      0.0    0.4200          0.490          Medium  2014   
51287        3      0.0   12.3600          0.350          Medium  2014   
51288        1      0.0    0.5600          0.199          Medium  2014   
51289        3      0.2   -0.6048          0.170          Medium  2014   

       category1 converted_order_date  
0          nilam           2011-01-01  
1          nilam           2011-01-01  
2          nilam           2011-01-01  
3          nilam           2011-01-01  
4          nilam           2011-01-01  
...          ...                  ...  
51285      nilam           2014-12-31  
51286      nilam           2014-12-31  
51287      nilam           2014-12-31  
51288     

In [90]:
df1['order_date_year'] = df1['converted_order_date'].dt.year

In [91]:
df1['order_date_month'] = df1['converted_order_date'].dt.month

In [92]:
df1['order_date_week'] = df1['converted_order_date'].dt.week

C:\Users\ACER\AppData\Local\Temp\ipykernel_9116\3626664048.py:1: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated. Please use Series.dt.isocalendar().week instead.
  df1['order_date_week'] = df1['converted_order_date'].dt.week


In [93]:
df1[['converted_order_date','order_date_week']]

converted_order_date  order_date_week
0               2011-01-01               52
1               2011-01-01               52
2               2011-01-01               52
3               2011-01-01               52
4               2011-01-01               52
...                    ...              ...
51285           2014-12-31                1
51286           2014-12-31                1
51287           2014-12-31                1
51288           2014-12-31                1
51289           2014-12-31                1

[51290 rows x 2 columns]

In [94]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1', 'converted_order_date', 'order_date_year',
       'order_date_month', 'order_date_week'],
      dtype='object')

In [95]:
df1[df1["order_date_year"] == 2014 ]

order_id order_date  ship_date       ship_mode  \
33759  ES-2014-2774938 2014-01-01 2014-01-03    Second Class   
33760    IN-2014-47134 2014-01-01 2014-01-05  Standard Class   
33761  IT-2014-4810734 2014-01-01 2014-01-03    Second Class   
33762   US-2014-105830 2014-01-01 2014-01-03     First Class   
33763   CA-2014-160395 2014-01-01 2014-01-07  Standard Class   
...                ...        ...        ...             ...   
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class   
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class   
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class   
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class   
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class   

          customer_name    segment             state        country  market  \
33759       Fred Harton   Consumer     Ile-de-France         France      EU   
33760       Kean Nguyen  Corporate          Victoria      Australia    APAC   
33761      Alan Shonely   Consumer     Ile-de-France         France      EU   
33762      Duane Benoit   Consumer              Ohio  United States      US   
33763     Kelly Lampkin  Corporate            Nevada  United States      US   
...                 ...        ...               ...            ...     ...   
51285        Erica Bern  Corporate        California  United States      US   
51286         Liz Preis   Consumer  Souss-Massa-Draâ        Morocco  Africa   
51287  Charlotte Melton   Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen   Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias   Consumer          Colorado  United States      US   

        region  ... discount    profit shipping_cost order_priority  year  \
33759  Central  ...     0.15 -111.6225       104.690           High  2014   
33760  Oceania  ...     0.10   46.9980        46.200         Medium  2014   
33761  Central  ...     0.15  -43.1370        37.850       Critical  2014   
33762     East  ...     0.20  -35.2152        30.660           High  2014   
33763     West  ...     0.20   53.9217        26.900            Low  2014   
...        ...  ...      ...       ...           ...            ...   ...   
51285     West  ...     0.20    4.5188         0.890         Medium  2014   
51286   Africa  ...     0.00    0.4200         0.490         Medium  2014   
51287  Central  ...     0.00   12.3600         0.350         Medium  2014   
51288    North  ...     0.00    0.5600         0.199         Medium  2014   
51289     West  ...     0.20   -0.6048         0.170         Medium  2014   

       category1  converted_order_date  order_date_year  order_date_month  \
33759      nilam            2014-01-01             2014                 1   
33760      nilam            2014-01-01             2014                 1   
33761      nilam            2014-01-01             2014                 1   
33762      nilam            2014-01-01             2014                 1   
33763      nilam            2014-01-01             2014                 1   
...          ...                   ...              ...               ...   
51285      nilam            2014-12-31             2014                12   
51286      nilam            2014-12-31             2014                12   
51287      nilam            2014-12-31             2014                12   
51288      nilam            2014-12-31             2014                12   
51289      nilam            2014-12-31             2014                12   

      order_date_week  
33759               1  
33760               1  
33761               1  
33762               1  
33763               1  
...               ...  
51285               1  
51286               1  
51287               1  
51288               1  
51289               1  

[17531 rows x 26 columns]

In [96]:
df1["order_date_month"].value_counts() #group by count, max sales in month of november and min sales in feb month

11    6342
12    6302
9     5975
6     5331
8     5078
10    4490
5     3747
7     3166
4     3057
3     3035
1     2599
2     2168
Name: order_date_month, dtype: int64

In [97]:
# expenditure by company discount + shipping_cost

In [98]:
df1["cost_company"] = df1["discount"] + df1["shipping_cost"]

In [99]:
df1["cost_company"]

0        35.460
1         9.820
2         8.170
3         5.320
4         4.800
          ...  
51285     1.090
51286     0.490
51287     0.350
51288     0.199
51289     0.370
Name: cost_company, Length: 51290, dtype: float64

In [100]:
df1

order_id order_date  ship_date       ship_mode  \
0         AG-2011-2040 2011-01-01 2011-01-06  Standard Class   
1        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
2         HU-2011-1220 2011-01-01 2011-01-05    Second Class   
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class   
4        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
...                ...        ...        ...             ...   
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class   
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class   
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class   
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class   
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draâ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ...    profit shipping_cost order_priority  year  category1  \
0       Africa  ...  106.1400        35.460         Medium  2011      nilam   
1      Oceania  ...   36.0360         9.720         Medium  2011      nilam   
2         EMEA  ...   29.6400         8.170           High  2011      nilam   
3        North  ...  -26.0550         4.820           High  2011      nilam   
4      Oceania  ...   37.7700         4.700         Medium  2011      nilam   
...        ...  ...       ...           ...            ...   ...        ...   
51285     West  ...    4.5188         0.890         Medium  2014      nilam   
51286   Africa  ...    0.4200         0.490         Medium  2014      nilam   
51287  Central  ...   12.3600         0.350         Medium  2014      nilam   
51288    North  ...    0.5600         0.199         Medium  2014      nilam   
51289     West  ...   -0.6048         0.170         Medium  2014      nilam   

       converted_order_date  order_date_year  order_date_month  \
0                2011-01-01             2011                 1   
1                2011-01-01             2011                 1   
2                2011-01-01             2011                 1   
3                2011-01-01             2011                 1   
4                2011-01-01             2011                 1   
...                     ...              ...               ...   
51285            2014-12-31             2014                12   
51286            2014-12-31             2014                12   
51287            2014-12-31             2014                12   
51288            2014-12-31             2014                12   
51289            2014-12-31             2014                12   

       order_date_week cost_company  
0                   52       35.460  
1                   52        9.820  
2                   52        8.170  
3                   52        5.320  
4                   52        4.800  
...                ...          ...  
51285                1        1.090  
51286                1        0.490  
51287                1        0.350  
51288                1        0.199  
51289                1        0.370  

[51290 rows x 27 columns]

In [101]:
df1["cost_company"].max()

933.57

In [102]:
df1["cost_company"].min()

0.002

In [103]:
df1[df1["cost_company"] == max(df1["cost_company"])]["customer_name"]

13891    Rick Hansen
Name: customer_name, dtype: object

In [104]:
df1[df1["cost_company"] == max(df1["cost_company"])]["product_name"]

13891    Plantronics CS510 - Over-the-Head monaural Wir...
Name: product_name, dtype: object

In [105]:
# deleting columns: axis = 0 (x axis) means row, axis = 1 (y axis) means column


In [106]:
df1.drop("cost_company" , axis =1) # it drops column from current instance but not from original dataframe

order_id order_date  ship_date       ship_mode  \
0         AG-2011-2040 2011-01-01 2011-01-06  Standard Class   
1        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
2         HU-2011-1220 2011-01-01 2011-01-05    Second Class   
3      IT-2011-3647632 2011-01-01 2011-01-05    Second Class   
4        IN-2011-47883 2011-01-01 2011-01-08  Standard Class   
...                ...        ...        ...             ...   
51285   CA-2014-115427 2014-12-31 2015-01-04  Standard Class   
51286     MO-2014-2560 2014-12-31 2015-01-05  Standard Class   
51287   MX-2014-110527 2014-12-31 2015-01-02    Second Class   
51288   MX-2014-114783 2014-12-31 2015-01-06  Standard Class   
51289   CA-2014-156720 2014-12-31 2015-01-04  Standard Class   

          customer_name      segment             state        country  market  \
0       Toby Braunhardt     Consumer       Constantine        Algeria  Africa   
1           Joseph Holt     Consumer   New South Wales      Australia    APAC   
2         Annie Thurman     Consumer          Budapest        Hungary    EMEA   
3          Eugene Moren  Home Office         Stockholm         Sweden      EU   
4           Joseph Holt     Consumer   New South Wales      Australia    APAC   
...                 ...          ...               ...            ...     ...   
51285        Erica Bern    Corporate        California  United States      US   
51286         Liz Preis     Consumer  Souss-Massa-Draâ        Morocco  Africa   
51287  Charlotte Melton     Consumer           Managua      Nicaragua   LATAM   
51288     Tamara Dahlen     Consumer         Chihuahua         Mexico   LATAM   
51289     Jill Matthias     Consumer          Colorado  United States      US   

        region  ... discount    profit shipping_cost order_priority  year  \
0       Africa  ...      0.0  106.1400        35.460         Medium  2011   
1      Oceania  ...      0.1   36.0360         9.720         Medium  2011   
2         EMEA  ...      0.0   29.6400         8.170           High  2011   
3        North  ...      0.5  -26.0550         4.820           High  2011   
4      Oceania  ...      0.1   37.7700         4.700         Medium  2011   
...        ...  ...      ...       ...           ...            ...   ...   
51285     West  ...      0.2    4.5188         0.890         Medium  2014   
51286   Africa  ...      0.0    0.4200         0.490         Medium  2014   
51287  Central  ...      0.0   12.3600         0.350         Medium  2014   
51288    North  ...      0.0    0.5600         0.199         Medium  2014   
51289     West  ...      0.2   -0.6048         0.170         Medium  2014   

       category1  converted_order_date  order_date_year  order_date_month  \
0          nilam            2011-01-01             2011                 1   
1          nilam            2011-01-01             2011                 1   
2          nilam            2011-01-01             2011                 1   
3          nilam            2011-01-01             2011                 1   
4          nilam            2011-01-01             2011                 1   
...          ...                   ...              ...               ...   
51285      nilam            2014-12-31             2014                12   
51286      nilam            2014-12-31             2014                12   
51287      nilam            2014-12-31             2014                12   
51288      nilam            2014-12-31             2014                12   
51289      nilam            2014-12-31             2014                12   

      order_date_week  
0                  52  
1                  52  
2                  52  
3                  52  
4                  52  
...               ...  
51285               1  
51286               1  
51287               1  
51288               1  
51289               1  

[51290 rows x 26 columns]

In [107]:
df1.columns

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_name',
       'segment', 'state', 'country', 'market', 'region', 'product_id',
       'category', 'sub_category', 'product_name', 'sales', 'quantity',
       'discount', 'profit', 'shipping_cost', 'order_priority', 'year',
       'category1', 'converted_order_date', 'order_date_year',
       'order_date_month', 'order_date_week', 'cost_company'],
      dtype='object')

# for permenant delete
1.df = df.drop('cost_company', axis= 1)

2.df.drop('cost_company', axis= 1, inplace= True)